benchmark/t3/helper.py

```diff
@@ -0,0 +1,122 @@
+import os
+import docker
+
+from benchmark.datasets import DATASETS, BigANNDataset
+
+def print_cuda_versions():
+    info = subprocess.check_output(['nvidia-smi', '--query-gpu=driver_version','--format=csv'])
+    print(info)
+    info = subprocess.check_output(['nvcc', '--version'])
+    print(info)
+
+
+def t3_create_container( definition, cmd, cpu_limit, mem_limit):
+
+    if definition.algorithm in [ 'faiss-t3' ]:
+
+        print("Launching GPU container")
+        container = create_container_with_gpu_support(
+            docker.from_env(),
+            definition.docker_tag,
+            cmd,
+            volumes={
+                os.path.abspath('benchmark'):
+                    {'bind': '/home/app/benchmark', 'mode': 'ro'},
+                os.path.abspath('data'):
+                    {'bind': '/home/app/data', 'mode': 'rw'},
+                os.path.abspath('results'):
+                    {'bind': '/home/app/results', 'mode': 'rw'},
+            },
+            cpuset_cpus=cpu_limit,
+            mem_limit=mem_limit,
+            detach=True)
+        container.start()
+        return container
+
+    else:
+        raise Exception("Docker invoke not supported for this algorithm.")
+
+def create_container_with_gpu_support(client, image, command, volumes, **kwargs):
+
+
+    from docker.models.images import Image
+    from docker.models.containers import _create_container_args
+
+    if isinstance(image, Image):
+        image = image.id
+
+    os.environ['NVIDIA_VISIBLE_DEVICES']='all'
+
+    kwargs['image'] = image
+    kwargs['command'] = command
+    kwargs['version'] = client.containers.client.api._version
+    kwargs['volumes'] = volumes
+    create_kwargs = _create_container_args(kwargs)
+
+    device_request = {
+        'Driver': 'nvidia',
+        'Capabilities': [['gpu'], ['nvidia']],
+        'Count': -1,  # enable all gpus
+    }
+
+    if device_request is not None:
+        create_kwargs['host_config']['DeviceRequests'] = [device_request]
+
+    resp = client.api.create_container(**create_kwargs)
+    return client.containers.get(resp['Id'])
```

Comment:
Awesome! I didn't know you could do that.

Reply:
Yes, in order to use the nvidia docker cuda run-time - you need to do some weird tricks.